lib/router/handler-info.js

```diff
@@ -36,22 +51,34 @@ HandlerInfo.prototype = {
     return this.params || {};
   },
 
+  updateHandler: function(handler) {
+    // Store the name of the handler on the handler for easy checks later
+    handler._handlerName = this.name;
+    return this.handler = handler;
+  },
+
   resolve: function(shouldContinue, payload) {
     var checkForAbort  = bind(this, this.checkForAbort,      shouldContinue),
         beforeModel    = bind(this, this.runBeforeModelHook, payload),
         model          = bind(this, this.getModel,           payload),
         afterModel     = bind(this, this.runAfterModelHook,  payload),
-        becomeResolved = bind(this, this.becomeResolved,     payload);
-
-    return Promise.resolve(undefined, this.promiseLabel("Start handler"))
-           .then(checkForAbort, null, this.promiseLabel("Check for abort"))
-           .then(beforeModel, null, this.promiseLabel("Before model"))
-           .then(checkForAbort, null, this.promiseLabel("Check if aborted during 'beforeModel' hook"))
-           .then(model, null, this.promiseLabel("Model"))
-           .then(checkForAbort, null, this.promiseLabel("Check if aborted in 'model' hook"))
-           .then(afterModel, null, this.promiseLabel("After model"))
-           .then(checkForAbort, null, this.promiseLabel("Check if aborted in 'afterModel' hook"))
-           .then(becomeResolved, null, this.promiseLabel("Become resolved"));
+        becomeResolved = bind(this, this.becomeResolved,     payload),
+        self = this;
+
+    return Promise.resolve(this.handlerPromise, this.promiseLabel("Start handler"))
+            .then(function(handler) {
+              return Promise.resolve(handler)
```

Comment:
Is this nesting needed? I could be missing something, but I would think this would work:

```js
return Promise.resolve(this.handlerPromise, this.promiseLabel("Start handler"))
    .then(checkForAbort, null, self.promiseLabel("Check for abort"))
    .then(beforeModel, null, self.promiseLabel("Before model"))
    .then(checkForAbort, null, self.promiseLabel("Check if aborted during 'beforeModel' hook"))
    .then(model, null, self.promiseLabel("Model"))
    .then(checkForAbort, null, self.promiseLabel("Check if aborted in 'model' hook"))
    .then(afterModel, null, self.promiseLabel("After model"))
    .then(checkForAbort, null, self.promiseLabel("Check if aborted in 'afterModel' hook"))
    .then(becomeResolved, null, self.promiseLabel("Become resolved"));
```